### PR TITLE
Partition full graph one execution provider at a time

### PR DESCRIPTION
**Pull request by skottmckay**
**Description**: 
Process the full graph, including subgraphs, with each execution provider instead of iterating through all EPs for each individual graph. 
Refactor to separate out partitioning and inlining better so we do them on a full graph basis instead of intertwining and calling Graph::Resolve at the subgraph level (which always runs Resolve against the full graph and not just the subgraph).

**Motivation and Context**
Simplifies how the partitioning is operating. 
Reduces number of calls to Graph::Resolve when partitioning.
Will simplify creating an ORT format model that supports multiple custom EPs as we will want to process the graph fully for each EP.

### CODE REVIEW

**Comment by KeDengMS** [comment]
How does it work for flow control ops? Should EP look at the subgraph inside the flow control ops to check if it can handle all of them, then take it over?

**Comment by skottmckay** [comment]
> How does it work for flow control ops? Should EP look at the subgraph inside the flow control ops to check if it can handle all of them, then take it over?

Yeah - one way or another. The idea of doing top-down was that it needed to do that anyway, so letting it see the control-flow node first saved processing cost for the subgraph if it takes it. However that cost is mitigated by doing the whole graph with each EP (vs. previously we ran all EPs on each subgraph before going up a level). To support nuphar I'm going to change it to do bottom-up, one EP at a time. I think that simplifies the check as well as the EP can simply see if all nodes in the subgraph are assigned to it when processing the control flow node.

File: onnxruntime/core/framework/graph_partitioner.h [hunk 1]
```diff
@@ -10,6 +10,7 @@
 
 namespace onnxruntime {
 
+class IExecutionProvider;
```
Comment (KeDengMS) [inline onnxruntime/core/framework/graph_partitioner.h:13]: nit: not needed?

**Review by KeDengMS — APPROVED** [review]
:shipit:

**Review by KeDengMS — APPROVED** [review]
:shipit:

**Pull request merged**